### PR TITLE
implemented service for booking status refreshing

### PR DESCRIPTION
**Pull request by DFSko**

### CODE REVIEW

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            return _recordManager.Get(bookingId)
+                .Bind(Refresh);
+
+            Task<Result> Refresh(Booking booking) => RefreshStatus(booking, userInfo, states);
+        }
+        
+        
+        private static readonly HashSet<BookingStatuses> BookingStatusesForRefresh = new()
+        {
+            BookingStatuses.Pending
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs:176]: Why here's only one status? All temporal and positively confirmed statuses must be checked time to time

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        };
+
+        private static readonly HashSet<BookingUpdateModes> BookingUpdateModesForRefresh = new()
+        {
+            BookingUpdateModes.Synchronous
+        };
+        
+        private static readonly Dictionary<int, TimeSpan> Delays = new()
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs:184]: `DelayStrategies` I guess

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            var states = await _flow.GetAsync<List<BookingRefreshStatusState>>(Key, Expiration);
+            
+            foreach (var bookingId in bookingIds)
+            {
+                var (_, isFailure, error) = await RefreshStatus(bookingId, userInfo, states);
+
+                if (isFailure)
+                {
+                    hasErrors = true;
+                    builder.AppendLine(error);
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs:132]: You may add negative results to `Result` directly

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                .Bind(CheckIsRefreshStatusNeeded)
+                .Bind(RefreshBookingStatus)
+                .Tap(CleanExpiredStates)
+                .Tap(UpdateStates);
+            
+            
+            Result ValidateBooking()
+            {
+                if (!BookingStatusesForRefresh.Contains(booking.Status))
+                    return Result.Failure($"Cannot refresh booking status for booking {booking.ReferenceCode}. Reason: {booking.Status} status is wrong");
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs:51]: 1. What means wrong?
2. Why not simply skip unsuitable statuses?

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        {
+            _flow = flow;
+            _dateTimeProvider = dateTimeProvider;
+            _bookingManagement = bookingManagement;
+            _context = context;
+            _recordManager = recordManager;
+        }
+
+
+        public async Task<Result> RefreshStatus(Booking booking, UserInfo userInfo, List<BookingRefreshStatusState> states = null)
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs:36]: Check of a single booking is a special case of a booking list check, so I advise to rewrite the service on top of `RefreshStatuses` method. That would be more efficient and save a lot of cycles

File: Api/Models/Bookings/BookingRefreshStatusState.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace HappyTravel.Edo.Api.Models.Bookings
+{
+    public record BookingRefreshStatusState
```
Comment (kirillta) [inline Api/Models/Bookings/BookingRefreshStatusState.cs:5]: `Status` and `State` mean the same, so it should be either `BookingRefreshState`, either `BookingStatusRefreshState`

**Review by kirillta — CHANGES_REQUESTED** [review]

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+using HappyTravel.Edo.Api.Models.Users;
+using HappyTravel.Edo.Common.Enums;
+using HappyTravel.Edo.Data;
+using HappyTravel.Edo.Data.Bookings;
+using HappyTravel.EdoContracts.Accommodations.Enums;
+using Microsoft.EntityFrameworkCore;
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Bookings.Management
+{
+    public class BookingRefreshStatusService : IBookingRefreshStatusService
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs]: Fixed

File: Api/Controllers/AgentControllers/InternalBookingsController.cs [hunk 8]
```diff
@@ -234,10 +238,37 @@ public async Task<IActionResult> MaterializeBookingsMarkupBonuses(List<int> appl
...
+        public async Task<IActionResult> GetBookingIdsForStatusRefresh()
+        {
+            return Ok(await _bookingRefreshStatusService.GetBookingsForUpdate());
+        }
+        
+        
+        /// <summary>
+        ///     Refresh booking statuses
+        /// </summary>
+        [HttpPost("refresh-booking-status")]
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Controllers/AgentControllers/InternalBookingsController.cs]: Fixed

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs (hunk 2 above)
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs]: Fixed

File: Api/Models/Bookings/BookingRefreshStatusState.cs (hunk 6 above)
Comment (DFSko) [reply Api/Models/Bookings/BookingRefreshStatusState.cs]: Fixed

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs (hunk 4 above)
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs]: Fixed texts

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs (hunk 5 above)
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs]: Fixed

File: Api/Models/Bookings/BookingStatusRefreshState.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace HappyTravel.Edo.Api.Models.Bookings
+{
+    public record BookingStatusRefreshState
+    {
+        public int BookingId { get; init; }
+        public int RefreshStatusCount { get; init; }
+        public DateTime LastRefreshingDate { get; init; }
```
Comment (mingazhev) [inline Api/Models/Bookings/BookingStatusRefreshState.cs:9]: "LastRefreshDate" I think will sound better

File: Api/Services/Accommodations/Bookings/Management/BookingStatusRefreshService.cs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+            return await _flow.GetAsync<List<BookingStatusRefreshState>>(Key, Expiration) 
+                ?? new List<BookingStatusRefreshState>();
+        }
+
+        private static readonly HashSet<BookingStatuses> BookingStatusesForRefresh = new()
+        {
+            BookingStatuses.Pending
+        };
+
+        private static readonly HashSet<BookingUpdateModes> BookingUpdateModesForRefresh = new()
```
Comment (mingazhev) [inline Api/Services/Accommodations/Bookings/Management/BookingStatusRefreshService.cs:187]: Should update not only synchronous bookings

File: Api/Services/Accommodations/Bookings/Management/IBookingRefreshStatusService.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Bookings.Management
+{
+    public interface IBookingRefreshStatusService
+    {
+        Task<Result> RefreshStatus(Booking booking, UserInfo userInfo, List<BookingRefreshStatusState> states = null);
+
+        Task<Result<BatchOperationResult>> RefreshStatuses(List<int> bookingIds, UserInfo userInfo);
+
+        Task<List<int>> GetBookingsForUpdate();
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/IBookingRefreshStatusService.cs]: Fixed

File: Api/Models/Bookings/BookingStatusRefreshState.cs (hunk 9 above)
Comment (DFSko) [reply Api/Models/Bookings/BookingStatusRefreshState.cs]: Fixed

File: Api/Services/Accommodations/Bookings/Management/BookingStatusRefreshService.cs (hunk 10 above)
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingStatusRefreshService.cs]: Fixed

File: Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+            Task<Result> Refresh(Booking booking) => RefreshStatus(booking, userInfo, states);
+        }
+        
+        
+        private static readonly HashSet<BookingStatuses> BookingStatusesForRefresh = new()
+        {
+            BookingStatuses.Pending
+        };
+
+        private static readonly HashSet<BookingUpdateModes> BookingUpdateModesForRefresh = new()
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Services/Accommodations/Bookings/Management/BookingRefreshStatusService.cs]: Fixed

**Review by mingazhev — APPROVED** [review]

**Review by kirillta — APPROVED** [review]

**Pull request merged**